src/sage/rings/species.py

```diff
@@ -0,0 +1,2390 @@
+            Traceback (most recent call last):
+            ...
+            ValueError: the assignment of sorts to the domain elements must be
+             provided
+
+            sage: M(G, {0:[1,2], 1:[]})
+            Traceback (most recent call last):
+            ...
+            ValueError: values of pi (=dict_values([[1, 2], []])) must
+             partition the domain of G (={1, 2, 3, 4, 5, 6})
+
+            sage: M(G, {1:[1,2], 2:[3,4,5,6]})
+            Traceback (most recent call last):
+            ...
+            ValueError: keys of pi (=dict_keys([1, 2])) must be in range(2)
+
+            sage: M(0)
+            Traceback (most recent call last):
+            ...
+            ValueError: 0 must be a permutation group or a pair specifying a
+             group action on the given domain pi=None
+        """
+        if parent(G) is self:
+            # pi cannot be None because of framework
+            raise ValueError("cannot reassign sorts to a molecular species")
+
+        if isinstance(G, PermutationGroup_generic):
+            if pi is None:
+                if self._arity == 1:
+                    pi = {0: G.domain()}
+                elif not G.degree():
+                    pi = {}
+                else:
+                    raise ValueError("the assignment of sorts to the domain elements must be provided")
+            elif not isinstance(pi, dict):
+                pi = {i: v for i, v in enumerate(pi)}
+            domain = [e for p in pi.values() for e in p]
+            if check and len(domain) != len(set(domain)) or set(G.domain()) != set(domain):
+                raise ValueError(f"values of pi (={pi.values()}) must partition the domain of G (={G.domain()})")
+            components = G.disjoint_direct_product_decomposition()
+            elm = self.one()
+            for component in components:
+                gens = [[cyc for cyc in gen.cycle_tuples() if cyc[0] in component]
+                        for gen in G.gens()]
+                H = PermutationGroup([gen for gen in gens if gen],
+                                     domain=component)
+                pi_H = {k: [e for e in v if e in component]
+                        for k, v in pi.items()}
+                a = self._indices(H, pi_H, check=check)
+                elm *= self.gen(a)
+            return elm
+
+        try:
+            X, a = G
+        except TypeError:
+            raise ValueError(f"{G} must be a permutation group or a pair specifying a group action on the given domain pi={pi}")
+        dompart = [sorted(pi.get(s, [])) for s in range(self._arity)]
+        S = PermutationGroup([tuple(b) for b in dompart if len(b) > 2]
+                             + [(b[0], b[1]) for b in dompart if len(b) > 1],
+                             domain=list(chain(*dompart)))
+        H = _stabilizer_subgroups(S, X, a)
+        if len(H) > 1:
+            raise ValueError("action is not transitive")
+        return self(H[0], pi, check=check)
```

Comment:
```suggestion
        return self._element_constructor_(H[0], pi, check=check)
```
This is where you expect it to always end up (without any coercion being done), right?

Reply:
Indeed.  I reorganized so that recursion is avoided, because it is a pain when debugging.

Also, the error message was (mathematically) wrong.